server/chat-plugins/mafia.ts

```diff
@@ -4057,23 +4126,161 @@ export const commands: Chat.ChatCommands = {
 			this.sendReply(`The entry ${entry} was deleted from the ${source} database.`);
 		},
 		deletedatahelp: [`/mafia deletedata source,entry - Removes an entry from the database. Requires % @ # ~`],
-		listdata(target, room, user) {
-			if (!(target in MafiaData)) {
-				throw new Chat.ErrorMessage(`Invalid source. Valid sources are ${Object.keys(MafiaData).join(', ')}`);
-			}
-			const dataSource = MafiaData[target as keyof MafiaData];
-			if (dataSource === MafiaData.aliases) {
-				const aliases = Object.entries(MafiaData.aliases)
-					.map(([from, to]) => `${from}: ${to}`)
-					.join('<br/>');
-				return this.sendReplyBox(`Mafia aliases:<br/>${aliases}`);
+
+		randtheme: 'listdata',
+		randrole: 'listdata',
+		randalignment: 'listdata',
+		randidea: 'listdata',
+		randterm: 'listdata',
+		randroles: 'listdata',
+		randalignments: 'listdata',
+		randideas: 'listdata',
+		randterms: 'listdata',
+		randdata: 'listdata',
+		randomtheme: 'listdata',
+		randomrole: 'listdata',
+		randomalignment: 'listdata',
+		randomidea: 'listdata',
+		randomterm: 'listdata',
+		randomdata: 'listdata',
+		randomthemes: 'listdata',
+		randomroles: 'listdata',
+		randomalignments: 'listdata',
+		randomideas: 'listdata',
+		randomterms: 'listdata',
+		randthemes: 'listdata',
+		listthemes: 'listdata',
+		listroles: 'listdata',
+		listalignments: 'listdata',
+		listideas: 'listdata',
+		listterms: 'listdata',
+		themes: 'listdata',
+		roles: 'listdata',
+		alignments: 'listdata',
+		ideas: 'listdata',
+		terms: 'listdata',
+		ds: 'listdata',
+		search: 'listdata',
+		random: 'listdata',
+		list: 'listdata',
+		listdata(target, room, user, connection, cmd, message) {
+			if (!this.runBroadcast()) return false;
+
+			// Determine non-search targets first, afterward searching is done with the remainder
+			const targets = target.split(',').map(x => x.trim().toLowerCase());
+
+			// Determine search type
+			let searchType: keyof MafiaData = 'aliases';
+			if (cmd.includes('theme') || targets.includes(`themes`)) {
+				searchType = `themes`;
+				if (targets.includes(`themes`)) targets.splice(targets.indexOf(`themes`), 1);
+			} else if (cmd.includes('role') || targets.includes(`roles`)) {
+				searchType = `roles`;
+				if (targets.includes(`roles`)) targets.splice(targets.indexOf(`roles`), 1);
+			} else if (cmd.includes('alignment') || targets.includes(`alignments`)) {
+				searchType = `alignments`;
+				if (targets.includes(`alignments`)) targets.splice(targets.indexOf(`alignments`), 1);
+			} else if (cmd.includes('idea') || targets.includes(`ideas`)) {
+				searchType = `IDEAs`;
+				if (targets.includes(`ideas`)) targets.splice(targets.indexOf(`ideas`), 1);
+			} else if (cmd.includes('term') || targets.includes(`terms`)) {
+				searchType = `terms`;
+				if (targets.includes(`terms`)) targets.splice(targets.indexOf(`terms`), 1);
+			} else if (targets.includes(`aliases`)) {
+				searchType = `aliases`;
```

Comment:
Lets apply D.R.Y. (Don't Repeat Yourself) to this.

Use a loop instead eg:
```suggestion
			const searchTypes = ['themes', 'roles', ...];
			for (const type of searchTypes) {
				if (cmd.includes(type) || targets.includes(type)) {
					searchType = type;
					if (targets.includes(type)) targets.splice(targets.indexOf(type), 1);
				}
			}
```

Don't use this exact suggestion as you'll need to adjust to handle random search types too. Any time you find yourself using the same code over and over with small adjustments, you should be looking for a way to use a loop instead (or similar).